### PR TITLE
Use new AGP 7 variant APIs + update to AGP 7.1

### LINKED ISSUE #90: Use new AGP variants API
Currently we use the now-legacy variants access APIs and these will be removed in AGP 8.0

### PR DESCRIPTION
**Pull request by ZacSweers**
Highlights:
* Use new AGP 7 variants API where possible
  * Only remaining usage is checking if a build type is minified
* Remove `variantFilter` API and use new `VariantBuilder.registerExtension` API to opt in variants, which turned out to be pretty sleek!
* Remove, effectively, all `afterEvaluate` calls (except one that we have for validating expected tasks)

Will plan to keep this un-merged until AGP 7.1 stable is out.

Resolves #90. Bit hacky in a couple spots and not everything we want/need is there, but we can work around them where needed.

### CODE REVIEW

**Comment by ZacSweers** [comment]
After this, just down to these internal APIs 👀 

```
import com.android.build.gradle.internal.publishing.AndroidArtifacts
import com.android.build.gradle.internal.publishing.AndroidArtifacts.ArtifactType
import com.android.build.gradle.internal.tasks.L8DexDesugarLibTask
import com.android.build.gradle.internal.tasks.R8Task
```

**Comment by ZacSweers** [comment]
Some of these might be possibly to extract to earlier PRs and run on AGP 7. Namely the extension to replace the variant filter and loosey goosey L8 configuration

**Comment by ZacSweers** [comment]
Nevermind, I don't think they can be easily detached as they both implicitly rely on using the new variant builder APIs, which we can't really do until we can leverage that CLASSES_DIR prop

File: docs/configuration.md [hunk 1]
```diff
@@ -1,6 +1,33 @@
...
+## Basic Configuration
+
+In order to enable Keeper on a particular variant, you must register the `KeeperVariantMarker`
+extension in AGP's `VariantBuilder` API like so:
+
+```kotlin
+androidComponents {
+  beforeVariants { builder ->
+    if (shouldRunKeeperOnVariant()) {
+      builder.optInToKeeper() // Helpful extension function
```
Comment (ZacSweers) [inline docs/configuration.md:10]: open to suggestions on this name

File: keeper-gradle-plugin/src/main/java/com/slack/keeper/KeeperPlugin.kt [hunk 2]
```diff
@@ -513,15 +483,33 @@ internal fun String.capitalize(locale: Locale): String {
...
 
-private class VariantFilterImpl(variant: BaseVariant) : VariantFilter {
-  @Suppress("PropertyName")
-  var _ignored: Boolean = false
+/**
+ * Similar to [TaskContainer.named], but waits until the task is registered if it doesn't exist,
+ * yet. If the task is never registered, then this method will throw an error after the
+ * configuration phase.
+ */
+private inline fun <reified T : Task> Project.namedLazy(
```
Comment (autonomousapps) [inline keeper-gradle-plugin/src/main/java/com/slack/keeper/KeeperPlugin.kt:491]: clever

**Review by autonomousapps — APPROVED** [review]

File: keeper-gradle-plugin/build.gradle.kts [hunk 3]
```diff
@@ -43,11 +42,16 @@ repositories {
 tasks.withType<KotlinCompile>().configureEach {
   kotlinOptions {
     jvmTarget = "1.8"
-    // Because Gradle's Kotlin handling is stupid
-    apiVersion = "1.4"
-    languageVersion = "1.4"
+    // Because Gradle's Kotlin handling is stupid, this falls out of date quickly
+    apiVersion = "1.5"
```
[the remark this answers is not in the document]
Comment (ZacSweers) [reply keeper-gradle-plugin/build.gradle.kts:46]: it has in the past but honestly who knows, best-efforting it

File: keeper-gradle-plugin/build.gradle.kts [hunk 4]
```diff
@@ -62,8 +66,13 @@ sourceSets {
 }
 
 java {
-  sourceCompatibility = JavaVersion.VERSION_1_8
-  targetCompatibility = JavaVersion.VERSION_1_8
+  toolchain {
+    languageVersion.set(JavaLanguageVersion.of(11))
```
[the remark this answers is not in the document]
Comment (ZacSweers) [reply keeper-gradle-plugin/build.gradle.kts:70]: yup!

File: keeper-gradle-plugin/build.gradle.kts [hunk 5]
```diff
@@ -62,8 +66,13 @@ sourceSets {
...
+  }
+}
+
+tasks.withType<JavaCompile>().configureEach {
+  options.release.set(8)
```
[the remark this answers is not in the document]
Comment (ZacSweers) [reply keeper-gradle-plugin/build.gradle.kts:75]: not language version, build with jdk 11 and target java 8 bytecode

File: keeper-gradle-plugin/src/main/java/com/slack/keeper/KeeperVariantMarker.kt [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+/**
+ * Register this with [VariantBuilder.registerExtension] to opt this variant into
+ * Keeper.
+ *
+ * ```
+ * androidComponents {
+ *   beforeVariants { variantBuilder ->
+ *     if (shouldRunKeeperOnVariant()) {
+ *       variantBuilder.registerExtension(KeeperVariantMarker.class, KeeperVariantMarker)
```
[the remark this answers is not in the document]
Comment (ZacSweers) [reply keeper-gradle-plugin/src/main/java/com/slack/keeper/KeeperVariantMarker.kt:13]: yeah just opting for making the class part explicit

File: sample/build.gradle [hunk 7]
```diff
@@ -86,20 +88,21 @@ boolean useTraceReferences = providers.gradleProperty("keeperTest.enableTraceRef
...
 
-keeper {
-  // Example: Only enable on "externalStaging"
-  variantFilter {
-    setIgnore(name != "externalStaging")
+// Example: Only enable on "externalStaging"
+androidComponents {
+  beforeVariants(selector().all()) { VariantBuilder variantBuilder ->
+    if (variantBuilder.name == "externalStaging") {
+      variantBuilder.registerExtension(KeeperVariantMarker.class, KeeperVariantMarker.INSTANCE)
```
[the remark this answers is not in the document]
Comment (ZacSweers) [reply sample/build.gradle:95]: groovy is a bastardization and I will use my strong types where possible :P

File: keeper-gradle-plugin/src/main/java/com/slack/keeper/KeeperPlugin.kt [hunk 8]
```diff
@@ -148,28 +146,37 @@ public class KeeperPlugin : Plugin<Project> {
...
+    ) { testVariant, appVariant ->
+      // TODO ideally move to components entirely https://issuetracker.google.com/issues/199411020
       if (appExtension.compileOptions.isCoreLibraryDesugaringEnabled) {
-        appExtension.onApplicableVariants(project, extension) { testVariant, appVariant ->
-
-          // First merge the L8 rules into the app's L8 task
-          val inputFiles = r8TaskFor(testVariant.name)
-              .flatMap { it.projectOutputKeepRules }
+        // namedLazy nesting here is unfortunate but necessary because these R8/L8 tasks don't
+        // exist yet during this callback. https://issuetracker.google.com/issues/199509581
```
[the remark this answers is not in the document]
Comment (ZacSweers) [reply keeper-gradle-plugin/src/main/java/com/slack/keeper/KeeperPlugin.kt:160]: they're not intentionally doing it, I have a FR to expose it

File: keeper-gradle-plugin/src/main/java/com/slack/keeper/KeeperPlugin.kt (hunk 2 above)
Comment (ZacSweers) [reply keeper-gradle-plugin/src/main/java/com/slack/keeper/KeeperPlugin.kt:491]: borrowed from @vRallev in anvil!

File: keeper-gradle-plugin/src/main/java/com/slack/keeper/KeeperPlugin.kt [hunk 9]
```diff
@@ -94,7 +94,7 @@ public class KeeperPlugin : Plugin<Project> {
     const val PRINTUSES_DEFAULT_VERSION = "1.6.53"
     const val TRACE_REFERENCES_DEFAULT_VERSION = "3.0.9-dev"
     const val CONFIGURATION_NAME = "keeperR8"
-    private val MIN_GRADLE_VERSION = GradleVersion.version("6.0")
+    private val MIN_GRADLE_VERSION = GradleVersion.version("7.2")
```
[the remark this answers is not in the document]
Comment (ZacSweers) [reply keeper-gradle-plugin/src/main/java/com/slack/keeper/KeeperPlugin.kt:97]: that's the minimum for AGP 7.1 https://developer.android.com/studio/releases/gradle-plugin#compatibility-7-1-0

**Pull request merged**